ai2thor/server.py

```diff
@@ -22,6 +23,263 @@ def default(self, obj):
+
+        detections = []
+
+        for color in self.instance_masks.class_colors.get(cls, []):
+            mask = self.instance_masks.instance_segmentation_frame_uint32 == self.instance_masks._integer_color_key(color)
+            bb = self.mask_bounding_box(mask)
+            if bb:
+                detections.append(bb)
+        if detections:
+            self._detections2d[cls] = detections
+        else:
+            raise KeyError(cls)
+
+        return self._detections2d[cls]
+
+class LazyMask(Mapping):
+
+    def __contains__(self, key):
+        return self.mask(key) is not None
+
+    def __getitem__(self, key):
+
+        m = self.mask(key)
+
+        if m is None:
+            raise KeyError(key)
+
+        return m
+
+    def __iter__(self):
+        self._load_all()
+        return iter(self._masks)
+
+    def __len__(self):
+        self._load_all()
+        return len(self._masks)
+
+
+
+class LazyInstanceSegmentationMasks(LazyMask):
+
+    def __init__(self, image_ids_data, metadata):
+        self._masks = {}
+        self._loaded = False
+        screen_width = metadata["screenWidth"]
+        screen_height = metadata["screenHeight"]
+        item_size = int(len(image_ids_data)/(screen_width * screen_height))
+        self._unique_integer_keys = None
+        self._empty_mask = None
+
+        if item_size == 4:
+            self.instance_segmentation_frame_uint32 = read_buffer_image(
+                image_ids_data, screen_width, screen_height, dtype=np.uint32
+            ).squeeze()
+            # within ImageSynthesis.cs a RGBA32 frame is read, the alpha channel
+            # is always equal to 255
+            self._alpha_channel_value = 255
+
+        elif item_size == 3: # 3 byte per pixel for backwards compatibility, RGB24 texture
+            # this is more expensive than the 4 byte variant since copying is required
+            frame = read_buffer_image(image_ids_data, screen_width, screen_height)
+            self.instance_segmentation_frame_uint32 = np.concatenate(
+                (
+                    frame,
+                    np.zeros((frame.shape[0], frame.shape[1], 1), dtype=np.uint8),
+                ),
+                axis=2,
+            )
+            self.instance_segmentation_frame_uint32.dtype = np.uint32
+            self.instance_segmentation_frame_uint32 = self.instance_segmentation_frame_uint32.squeeze()
+            self._alpha_channel_value = 0
+
+        # At this point we should have a 2d matrix of shape (height, width)
+        # with a 32bit uint as the value
+
+        self.instance_colors = {}
+        self.class_colors = {}
+        for c in metadata["colors"]:
+            cls = c["name"]
+
+            if "|" in c["name"]:
+                self.instance_colors[c["name"]] = c["color"]
+                cls = c["name"].split("|")[0]
+
+            if cls not in self.class_colors:
+                self.class_colors[cls] = []
+
+            self.class_colors[cls].append(c["color"])
```

Comment:
This seems to make the assumption that a object's name will always be of the form `OBJECT_TYPE|SOMETHING`, I don't think that's generally true (especially in nanna). I think we'll need to cross reference this against `metadata["objects"]` to get the class.

Reply:
Ah, looking below I suppose this is for backwards compatibility reasons? I think this is something we should think about breaking as it's a bit weird to have things like `StandardUpperCabinetHeightWidth` as a class.

Reply:
Yes, exactly.  This PR should have the same class/objectId entries as the previous version.  For reference, this is where the class names come from: https://github.com/allenai/ai2thor/blob/main/unity/Assets/Scripts/ImageSynthesis/ImageSynthesis.cs#L300